WIP: ENH: fft: support array API

lucascolley opened the pull request:
### Reference issue
RFC for array API support in SciPy: https://github.com/scipy/scipy/issues/18286
Tracker for array API support in SciPy: https://github.com/scipy/scipy/issues/18867
Merged PR adding machinery and covering `scipy.cluster`: https://github.com/scipy/scipy/pull/18668

(This PR follows on from issue https://github.com/tupui/scipy/issues/26)

### What does this implement/fix?
This PR extends the array API support of https://github.com/scipy/scipy/pull/18668 to also cover `scipy.fft`.

`fft` is an [array API standard extension module](https://data-apis.org/array-api/latest/extensions/fourier_transform_functions.html) and has matching APIs in both CuPy and PyTorch. The goal is to fully cover `scipy.fft` to conform to the array API standard, as specified in [the array API standard documentation](https://data-apis.org/array-api/latest/purpose_and_scope.html#conformance):

> A conforming implementation of the array API standard must provide and support all the functions, arguments, data types, syntax, and semantics described in this specification.
> 
> A conforming implementation of the array API standard may provide additional values, objects, properties, data types, and functions beyond those described in this specification.

For more context on the standard and the goal of SciPy adoption, please see [the RFC](https://github.com/scipy/scipy/issues/18286).

For explanation of the machinery of which this PR makes use, please see [the `cluster` PR](https://github.com/scipy/scipy/pull/18668).

**Changes:**
- The `uarray` dispatch structure of `_basic.py` has been moved to a new file, `_basic_uarray.py`, which is used for NumPy arrays and array-likes. All input arrays are now validated by `array_namespace`, so all valid input arrays either support the array API standard or are converted to NumPy arrays. The functions of `_basic.py` fall into two categories in their handling of non-NumPy arrays:
1. Those which are part of the standard first raise exceptions if unsupported parameters are provided. They then check whether the array's namespace has an `fft` module. If it does, then that is used. If not then a conversion to NumPy is attempted in order to use `uarray`. If successful, the result is converted back to the same namespace as the input array. For an example, see [`fft.fft`](https://github.com/lucascolley/scipy/blob/a097d91fa0fa569f167a18610c668ed4acfc07ba/scipy/fft/_basic.py#L10).
2. For those which are not part of the standard, a conversion to NumPy is attempted in order to use `uarray`. If successful, the result is converted back to the same namespace as the input array. For an example, see [`fft.fft2`](https://github.com/lucascolley/scipy/blob/a097d91fa0fa569f167a18610c668ed4acfc07ba/scipy/fft/_basic.py#L942). This means that GPU support is restricted to functions which are part of the standard for now, as otherwise they are hitting compiled code made for NumPy arrays. This design decision is explained in the RFC:
<details>

<summary>RFC excerpt (expandable)</summary>

> When a non-NumPy array type sees compiled code in SciPy (which tends to use the NumPy C API), we have a couple of options:
> 
> 1. dispatch back to the other library (PyTorch, CuPy, etc.).
> 
> 2. convert to a NumPy array when possible (i.e., on CPU via the buffer protocol, DLPack, or __array__), use the compiled code in question, then convert back.
> 
> 3. don't support this at all
> 
> I'll note that (1) is the long-term goal; how to implement this is out of scope for this proposal - for more on that, see the Appendix section. For now we choose to do (2) when possible, and (3) otherwise. Switching from that approach to (1) in the future will be backwards-compatible.


</details>

- The same design has been applied to `_realtransforms.py` and `_realtransforms_uarray.py`, save for the fact that there are no standard functions in these files.
- `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift` were previously just imported from `numpy`. They now have implementations in `_helper.py`, and follow the pattern for standard functions.
- The NumPy backend of `_fftlog.py` has been moved to a new file, `_fftlog_np.py`. Its `uarray` dispatch structure remains in the file `_fftlog_multimethods.py`. The functions now follow the pattern for non-standard functions.
- Tests for `fftfreq` etc. have been copied over from NumPy to `test_helper.py`.
- Tests have been modified to be array API compatible where possible.
- `TestNamespaces` has been added to `test_basic.py` and `test_helper.py` to check that output arrays are of the same namespace as input arrays.
- `test_non_standard_params` has been added to `test_basic.py` to check that exceptions are raised for unsupported parameters.
- A new helper has been added to `conftest.py`: `set_assert_allclose(xp)`, which returns an `assert_allclose` equivalent for the namespace `xp` (currently supports `cupy` and `torch`), to allow testing on GPU.
- Two new helpers have been added to `_array_api.py`: `is_numpy(xp)` which checks whether `xp.__name__` is `"scipy._lib.array_api_compat.array_api_compat.numpy"`, and `_assert_matching_namespace` by Tyler Reddy.

### Additional information

**To-do:**
- https://github.com/lucascolley/scipy/blob/f5bc71ab4adf33bd7b35fb95e1c422320ac70a3e/scipy/fft/tests/test_fftlog.py#L118-L122 (written by https://github.com/ntessore).

**Difficulties:**
- Some errors have come up where the floating point precision hasn't quite been good enough. If somebody could check that any introductions/modifications of `rtol` are appropriate, that would be greatly appreciated.
- `test_basic.py::test_multiprocess` is currently skipped for array API since it hangs for any array API library (even `numpy.array_api`). Generally, multithreading tests (including those using the unsupported `workers` keyword in `test_multithreading.py` are skipped apart from `TestFFTThreadSafe` in `test_basic.py`.
- `@skip_if_array_api_backend` and `@pytest.mark.parametrize` don't seem to work together, raising exceptions for the backends which aren't skipped, no matter which decorator comes first. If anyone could get this working, that would be great!
- `@skip_if_array_api_gpu` currently skips on every backend when `python dev.py test -b all` is used, due to a fault in its logic. This is not critical since the tests can be run separately for each backend, but this is unintended behaviour which should be fixed.

**Implementation Details:**
- Any tests of non-standard functions (which involve attempted conversion to NumPy upon reaching compiled code) are skipped on GPU backends.
- The signatures for `fftfreq` and `rfftfreq` have changed to `(n, d=1.0, *, xp=None, device=None)`. The addition of `device` is to match the signature in the standard. The addition of `xp` is to specify the namespace for the output array. This is required since these are "array-generating functions" which output an array without taking one as input. Since [array-api-compat](https://github.com/data-apis/array-api-compat) and [`numpy.array_api`](https://numpy.org/devdocs/reference/array_api.html) are yet to implement `fft`, these functions and their tests currently require workarounds/`pytest.xfail`'s. A lot of cleanup will be possible at a later date. https://github.com/lucascolley/scipy/blob/2cb6268ffde40a1685a803f9da37af24de80883d/scipy/fft/_helper.py#L147-L152 https://github.com/lucascolley/scipy/blob/a097d91fa0fa569f167a18610c668ed4acfc07ba/scipy/fft/tests/test_helper.py#L444-L449 These implementations will close https://github.com/scipy/scipy/issues/13887.
Perhaps a warning would be good to state that a NumPy array is returned unless the `xp` parameter is provided?
- `test_basic.py::test_fft_with_order` is skipped since orders are not supported in the array API. Likewise, `test_identity_1d_overwrite` and `test_identity_nd_overwrite` in `test_real_transforms.py` are skipped since the `overwrite_x` keyword is not supported.
- Some of the tests in `test_basic.py::TestFFT1D` and `test_helper.py::TestFFTShift` are failing on `torch` backend due to differences in the function signatures, but they will pass once array-api-compat has implemented `torch.fft`. Currently either `@skip_if_array_api_backend('torch')` or `pytest.xfail` is used.
- Justification for why we have `set_assert_allclose` instead of just converting to NumPy and using `assert_allclose`: https://github.com/data-apis/array-api/issues/626. Additional dtype casts are needed for `torch.assert_close` e.g. in `test_float32` (a part of the previous `test_dtypes`) and in `test_basic.py::TestFFTThreadSafe`.

**Follow-up:**
- https://github.com/lucascolley/scipy/blob/5f454109dbac63b1ff7a2b6a191ae547db6e382e/scipy/fft/tests/test_basic.py#L583-L586
- Docstrings need to be updated for the array API. This includes the new keywords for `fftfreq` and `rfftfreq`.
- The `device` keyword of `fftfreq` and `rfftfreq` are not currently tested since the keyword only works with PyTorch currently. But this should probably be tested once more array libraries support it.
- We may want to rename `_fftlog_multimethods.py` to `_fftlog_uarray.py`. Currently omitted to keep the diff clean.
- Functions outside of the array API like `fft2` could be rewritten in terms of the standard functions in order to avoid conversion to NumPy, thus allowing them to run on GPU.

Many thanks to Pamphile, Irwin and Ralf who have given really helpful guidance in the making of this PR!

**References:**
- Array API Standard: https://data-apis.org/array-api/latest/index.html
- Standard FFT Page: https://data-apis.org/array-api/latest/extensions/fourier_transform_functions.html
- Current SciPy FFT API: https://docs.scipy.org/doc/scipy/reference/fft.html
- `numpy.array_api`: https://numpy.org/devdocs/reference/array_api.html
- array-api-compat: https://github.com/data-apis/array-api-compat

lucascolley commented:
cc: @tupui @izaid

**Current next big step:**
- Tests: modify the tests for a few functions to make them array API compatible, and get them passing.


**To-do list:**
- Throw the correct `ValueError`'s instead of the current `Exception`'s
- Ensure the use of `*`'s in `fftfreq` and `rfftfreq` (to which they have been added) is correct and ensure that their uses are correct in all other functions (since the array API does not match the SciPy API).
- Check whether the current use of `isinstance` is correct; I have a feeling it may not account for all types of NumPy array. Perhaps use `array_namespace` first before checking whether it is a NumPy array; developing a new `isNumpy` function has been suggested
- `fftlog`: I couldn't get this working when I tried to modify the functions like those in `_realtransforms.py` since the implementation with the `uarray` multimethods is complicated. Try again and get help if still stuck.
- Look at all of the non array API functions and rewrite them to support different namespaces where possible - perhaps just when they are pure python code - rather than only attempting to convert to NumPy.

izaid commented on scipy/fft/_basic.py:
```diff
@@ -1,1629 +1,381 @@
+        return xp.fft.fft(x, n=n, axis=axis, norm=norm)
+    x = np.asarray(x)
+    y = npbasic.fft(x, n=n, axis=axis, norm=norm)
+    return xp.asarray(y)
```
FFT (and therefore all the similar functions) look good!

izaid commented on scipy/fft/_basic.py:
I'm happy if we leave FFT2 and IFFT2 like this right now, but we should do update them in a follow-up PR.

izaid commented on scipy/fft/_helper.py:
```diff
@@ -1,101 +1,70 @@
+    # device is currently not supported in numpy, cupy or array-api-compat
+    if device is not None:
+        raise ValueError(
+            "Providing the 'device' parameter is not yet supported.")
```
Is this a mistake? I'm not sure. Shouldn't we just forward the device parameter to the xp fftfreq? But if device is not None and xp == np, it should certainly raise an error.

lucascolley replied:
Forwarding the device parameter to `xp.fft.fftfreq` would work for _some_ libraries right now e.g. PyTorch. However, it won't work with CuPy, as - like NumPy - there is no `device` parameter currently. We could definitely just pass to `xp` and let CuPy throw an exception, but my reasoning for not doing this yet is to not 'jump ahead' of array-api-compat. Since array-api-compat already has helpers like `to_device`, I think that this functionality will be added once they get around to implementing `fft` (they already have `linalg`).

In fact, once array-api-compat has implemented `fft`, I think that there will be no issue with `xp=np` and `device` being not None, since we can leave array-api-compat to raise the exceptions:
```
if _is_numpy_array(x):
        if stream is not None:
            raise ValueError("The stream argument to to_device() is not supported")
        if device == 'cpu':
            return x
        raise ValueError(f"Unsupported device {device!r}")
```
(excerpt from `to_device` in array-api-compat)

izaid replied:
Indeed, you are totally correct. CuPy doesn't even support the `device` keyword in `fftfreq` at the moment.

I think we should do what you said, and pass it in anyway and let CuPy throw the exception. It also means we are doing the "right thing", and signals to the array libraries what exactly they need to support.

tupui replied:
Yes if we add parameters to comply with the standard, it's out of our hands. We only need to deal with SciPy handling it or not. e.g. if we go to compiled code instead.

tupui replied:
Sorry I am not following. If users pass in a np array, then the output is a np array no? Why a warning?

lucascolley replied:
No array is passed in! This is the problem and the whole need for the `xp` parameter.

tupui replied:
Ah yes I see. Sorry I was not paying enough attention to the surrounding code.

TBH in such cases I would not touch the functions if we don't do anything than returning the NumPy one. In the end this is something we would want to remove from SciPy. I would leave this out of this work and make a note in the PR description. People working on the module would have an opinion and I would not want you to do too much work for nothing.

lucascolley replied:
I saw this the other day saying why we would want to leave this in the SciPy API. Definitely something to think about but maybe to leave as a comment for a follow-up PR.
https://github.com/scipy/scipy/issues/13887#issuecomment-821987280

lucascolley replied:
Should I change it to just always returning the NumPy result for now? If so, then again I think it would be useful to provide a warning since this would be unexpected for someone using CuPy.

tupui replied:
Good find 👍 The discussion is still up so for now we have +2 and -1. We can link to that in the description of the PR.

lucascolley replied:
I would be inclined to agree with the -1 since the NumPy docs say 

> The SciPy module [scipy.fft](https://docs.scipy.org/doc/scipy/reference/fft.html#module-scipy.fft) is a more comprehensive superset of numpy.fft, which includes only a basic set of routines.

It seems to me that breaking this superset relationship might cause more issues than it is worth, but that's a decision for you guys to make!

lucascolley replied:
I'll resolve this conversation once we've decided what to do with these functions.

izaid commented on scipy/fft/tests/test_basic.py:
```diff
@@ -5,52 +5,81 @@
 
 class TestFFTShift:
 
+    @skip_if_array_api
```
Shouldn't we include this in the array api tests?

lucascolley replied:
Pushed a fix for this.

izaid commented on scipy/_lib/_testutils.py:
```diff
@@ -255,3 +257,14 @@ def _get_mem_available():
     return None
+
+
+def _assert_matching_namespace(actual, expected):
```
Thinking about this more, I do think we'd still want to assert that the return array of fftfreq is the same array as in `xp`. What I was trying to get at is the `_assert_matching_namespace` function won't work for the array creation functions that need to take in `xp` and `device`. In that case, we'll need something more like `_assert_matching_xp_and_device(actual, expected_xp, expected_device)` - but with a better name than that!

lucascolley replied:
Yes I agree, that's what I had in mind. It shouldn't be difficult to implement this since array-api-compat has a `check_device` helper. I can get this working for now, I'm just hesitant to be too set on this since `xp=np, device=None` is new and a different design could be chosen by the SciPy maintainers.

lucascolley replied:
Moved to `_array_api.py` and changed to `xp=None`.

tupui reviewed (commented):
Thanks @lucascolley 🚀 This is moving good.

I am thinking about the real PR you are going to open. I anticipate that the main critic you will get will be around the refactoring (move/rename) part. Ideally, the move part would be done in a single commit, the first one. This way we can tell maintainers to review this commit separately from the rest (in GitHub you can filter the commits you are looking at and here it will greatly help.)

lucascolley commented:
> Ideally, the move part would be done in a single commit, the first one.
Sounds good, how should I do this @tupui ?

tupui commented on scipy/fft/tests/test_basic.py:
```diff
@@ -5,52 +5,81 @@
 from numpy.testing import (
-        assert_array_almost_equal, assert_array_equal, assert_allclose
-        )
+    assert_array_almost_equal, assert_array_equal, assert_allclose
+)
```
Yes I am afraid it's the thing to do. A consolation is that we have a pre-commit hook you can install which does what the CI linter does. https://scipy.github.io/devdocs/dev/contributor/pep8.html

tupui commented:
> > Ideally, the move part would be done in a single commit, the first one.
> 
> Sounds good, how should I do this @tupui ?

If you did that, and just that, in a selection of commits, you would be able to squash these together. And then move the resulting commit to the beginning. That's tricky to do from the command line...

If not, then I am afraid there is not much choice (at least would be faster) than reseting the history of the branch and start fresh.

lucascolley commented:
> If you did that, and just that, in a selection of commits, you would be able to squash these together.

Could you have a look at the first commit on this branch? It doesn't have too much more than the move/rename part, but it is quite different from the current tip of the branch.

tupui commented:
> If you did that, and just that, in a selection of commits, you would be able to squash these together.
> 
> Could you have a look at the first commit on this branch? It doesn't have too much more than the move/rename part, but it is quite different from the current tip of the branch.

It's almost perfect 👍 Ideally this could be split in 2: 1 move the files, 2 add back the old files with just the new content. But ok let's keep this in mind for later if someone complains (or if you are looking for something to do 😅)

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -5,52 +5,81 @@
+            import cupy as cp
+            _assert_allclose = cp.testing.assert_allclose
+        else:
+            _assert_allclose = assert_allclose
```
Okay, just created `set_assert_allclose` in `conftest.py`. Let me know if that looks okay.

lucascolley commented:
thanks for the comments @tupui @izaid ! I think that I have addressed all of the comments now but please let me know if you think I've missed something. I've updated the to-do list accordingly :)

tupui commented:
One small thing, can you actually let the `Array API` and possibly just another job run on the CI? You need to edit the files in `.github/workflows` and comment out the line: `if: "github.repository == 'scipy/scipy' || github.repository == ''"`

lucascolley commented:
Array API CI successful, but `Linux Meson tests / Meson build (3.12-dev, false)` failing? The error certainly doesn't look like it would have been caused by me...
```
../scipy/integrate/_odepackmodule.c:477:12: error: implicit declaration of function ‘PyArray_MAX’; did you mean ‘PyArray_Max’? [-Werror=implicit-function-declaration]
  477 |     *lrw = PyArray_MAX(lrn,lrs);
      |            ^~~~~~~~~~~
      |            PyArray_Max
```

tupui commented:
This is strange but I also think it is not related to your PR. It's hard to get a green CI in SciPy 😓 but it's not a blocker to get something merge. There are many parameters changing constantly and recently it's been red on a few job. This one is fairly new so there might be things we still need to adjust. Sometimes just waiting a bit and re-running a test will make it pass.

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -0,0 +1,513 @@
+        for i in range(1, maxlen):
+            _assert_allclose(fft.ifft(fft.fft(x[0:i])), x[0:i], rtol=1e-12, atol=0)
+            _assert_allclose(
+                fft.irfft(fft.rfft(xr[0:i]), i), xr[0:i], rtol=1e-12, atol=0)
```
These were previously `assert_array_almost_equal` with `decimal=12`. They usually pass, but I have seen them occasionally fail to meet the precision required by `rtol=1e-12`  - by 1 element out of hundreds - on the `numpy` backend.

rgommers commented:
> I am thinking about the real PR you are going to open. I anticipate that the main critic you will get will be around the refactoring (move/rename) part. Ideally, the move part would be done in a single commit, the first one. This way we can tell maintainers to review this commit separately from the rest (in GitHub you can filter the commits you are looking at and here it will greatly help.)

I am likely missing a reason and/or previous discussion here, but: why are these large code moves / file renames necessary? It seems to me like it's splitting out the `uarray` dispatching code, however the docstrings for example should not be moved as far as I can tell. E.g. for the `fft` function:
- it's imported in `__init__.py` from `._basic` (not `._basic_np`),
- hence the docstring for the function should remain in `_basic.py`,
- the `For non-numpy arrays, this  ...` docstring is more a code comment than a docstring,
- `_np_basic.fft` returns a dispatchable, that's actually deserving of code comments (maybe at the top of `_basic_np.py`?) - it hits `_ScipyBackend` in `_backend.py` which actually forwards to the `fft._pocketfft` extension module which contains the actual implementation
- so once the docstrings stay, perhaps the `_basic_np.py` file should be named `_basic_uarray.py` instead, since all it will then contain is a thin layer of uarray dispatchers.

To verify the docstring behavior:
```
$ python dev.py ipython
>>> from scipy import fft
>>> fft.fft?
```

Also a minor comment on code style: try to not do `import ... as ...` unless you can avoid it. E.g., in `from . import _basic_np as npbasic` just remove the `as npbasic` and the code will be easier to read.

rgommers commented:
@lucascolley in every `*_uarray.py` file, I suggest putting the following comment right below the `import` statements:
```python
# The functions in this file decorated with `@_dispatch` will only be called
# from the correspondingly named public functions with NumPy arrays or
# array-likes, not with CuPy, PyTorch and other array API standard supporting
# objects. See the docstrings in `_backend.py` for more details and examples.
```

Let's leave out the longer comment in `_backend.py` for now, the examples are fairly clear there, so this should already help enough.

lucascolley commented:
Closing now as the full PR has been opened on the main repo: https://github.com/scipy/scipy/pull/19005
